src/prefect/task_engine.py

```diff
@@ -254,6 +216,106 @@ def _wait_for_dependencies(self):
             context={"current_task_run": self.task_run, "current_task": self.task},
         )
 
+    def record_terminal_state_timing(self, state: State) -> None:
+        if PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION:
+            if self.task_run.start_time and not self.task_run.end_time:
+                self.task_run.end_time = state.timestamp
+
+                if self.task_run.state.is_running():
+                    self.task_run.total_run_time += (
+                        state.timestamp - self.task_run.state.timestamp
+                    )
+
+    def is_running(self) -> bool:
+        """Whether or not the engine is currently running a task."""
+        if (task_run := getattr(self, "task_run", None)) is None:
+            return False
+        return task_run.state.is_running() or task_run.state.is_scheduled()
+
+    def log_finished_message(self):
+        # If debugging, use the more complete `repr` than the usual `str` description
+        display_state = repr(self.state) if PREFECT_DEBUG_MODE else str(self.state)
+        level = logging.INFO if self.state.is_completed() else logging.ERROR
+        msg = f"Finished in state {display_state}"
+        if self.state.is_pending():
+            msg += (
+                "\nPlease wait for all submitted tasks to complete"
+                " before exiting your flow by calling `.wait()` on the "
+                "`PrefectFuture` returned from your `.submit()` calls."
+            )
+            msg += dedent(
+                """
+
+                Example:
+
+                from prefect import flow, task
+
+                @task
+                def say_hello(name):
+                    print f"Hello, {name}!"
+
+                @flow
+                def example_flow():
+                    future = say_hello.submit(name="Marvin)
+                    future.wait()
+
+                example_flow()
+                            """
+            )
+        self.logger.log(
+            level=level,
+            msg=msg,
+        )
+
+
+@dataclass
+class SyncTaskRunEngine(BaseTaskRunEngine[P, R]):
+    _client: Optional[SyncPrefectClient] = None
+
+    @property
+    def client(self) -> SyncPrefectClient:
+        if not self._is_started or self._client is None:
+            raise RuntimeError("Engine has not started.")
+        return self._client
+
+    def sleep(self, interval: float):
```

Comment:
what's the idea with encapsulating this? It seems that the uses of it still require a switch between sync/async call patterns so it can't be generalized to a base pattern

Reply:
Yeah, it's really just for testing unfortunately. We use anyio.sleep for other things in here (like the wait for start method) and to test the task polling the test is mocking out sleep so this was done to let us mock that directly without the side-effects.

Reply:
For the test where I see this method being mocked (`test_paused_task_polling`), I don't think the `anyio.sleep` in `wait_until_ready` will run because there aren't any task retries configured. If I'm mistaken, and `anyio.sleep` does get called, I think it might be preferable to mock out `wait_until_ready` instead so that we don't need to add this method. I was a little confused that we use `self.sleep` in one place, but `anyio.sleep` in another, so I think it'd be good to consolidate on one way of sleeping for each engine implementation.

Reply:
Hmm, maybe it wasn't the wait method, I know I had trouble with just mocking anyio.sleep directly, but I can try it again and see if I can come up with a better approach.